pysmt/optimization/optimizer.py

```diff
@@ -151,49 +284,118 @@ def optimize(self, cost_function, strategy='ub',
 
         `step_size` the minimum reolution for finding a solution. The
         optimum will be found in the proximity of `step_size`
-
         """
+        rt = None, None
+        if goal.is_maximization_goal() or goal.is_minimization_goal():
+            rt = self._optimize(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        else:
+            raise GoalNotSupportedError("ExternalOptimizerMixin", goal)
+        return rt
+
+
+    def _optimize(self, goal, strategy,
+                 feasible_solution_callback,
+                 step_size, **kwargs):
+
+        current = OptSearchInterval(goal, self._comparation_functions)
+        first_step = True
+        while not current.empty():
+            if not first_step:
+                if strategy == "linear":
+                    lin_assertions = current.linear_search_cut(step_size)
+                elif strategy == "binary":
+                    lin_assertions = current.binary_search_cut()
+                else:
+                    raise PysmtValueError("Unknown optimization strategy '%s'" % strategy)
+                status = self.solve(assumptions=[lin_assertions])
```

Comment:
If you want to maintain the distinction between SUAOptimizers and Incremental optimizers, this part needs to be delegated to a mixin function